Detect unconditional redirects and ignore any rules defined afterwards

tomfotherby opened the pull request:
(_this is my first PR for GoLang code and for Kubernetes-related projects, so bear with me if I make mistakes or faux pas_)

ALBs only support 100 rules and we have hit the limit in our project -  we are looking for a way to reduce rules and have identified that the aws alb controller is creating some unneeded ALB rules.

We use a annotation to redirect port 80 requests to port 443. This is the annotation documentation: https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/docs/guide/tasks/ssl_redirect.md

If a unconditional redirect rule is present in the configuration then all following rules within the same listener are not needed. This PR changes things so they are skipped thereby conserving rules.

For the biggest benefit, it is best to define the redirect rule first in the config. Because only subsequent rules are skipped. This is in line with the ssl-redirect annotation which [says](https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/docs/guide/tasks/ssl_redirect.md):
> the ssl-redirect action must be be first rule(which will be evaluated first by ALB)

---

I have tested on our project, with success. But please let me know if I can make any improvements.

If this is not something that fits with the direction of the project, no problem, please close this PR. I only created it in case it's useful - I'm 100% happy to close if not the case.

---

Relvant GitHub issue https://github.com/kubernetes-sigs/aws-alb-ingress-controller/issues/853#issuecomment-578547468

tomfotherby commented:
/assign @bigkraig

mogren commented on internal/alb/ls/rules.go:
```diff
@@ -529,3 +537,53 @@ func createsRedirectLoop(listener *elbv2.Listener, r elbv2.Rule) bool {
+            case conditions.FieldHTTPHeader:
+                return false
+            case conditions.FieldQueryString:
+                return false
```
In go you could use a comma separated list for this:
```go
case conditions.FieldHTTPRequestMethod, conditions.FieldSourceIP, conditions.FieldHTTPHeader, conditions.FieldQueryString:
    return false
```

tomfotherby replied:
Thank you @mogren - I pushed a update with your suggestion included.

tomfotherby commented:
We have been running a ALB ingress controller that uses the code in this PR at https://peopleperhour.com in production successfully. It freed up > 40 ALB rules for us. The ALB gets about 110 requests per second:

![Selection_311](https://user-images.githubusercontent.com/140179/74457639-75f51d80-4e80-11ea-9aed-a2717e7a951f.png)

M00nF1sh commented:
/ok-to-test

M00nF1sh commented on internal/alb/ls/rules.go:
```diff
@@ -529,3 +537,48 @@ func createsRedirectLoop(listener *elbv2.Listener, r elbv2.Rule) bool {
+            return true
+        }
+        if len(paths) != 0 {
+            wildcardPathMatch := false
```
can be simplified to  🤣 
```
for _, path := range paths {
    if path == "/*" {
          wildcardPathMatch = true
          return true
    }
 }
```

tomfotherby replied:
How did I miss that! Thanks! I'll fix.

M00nF1sh commented:
would you help fix the lint..i'll clean up some vpcs in our test account to make e2e pass :D

tomfotherby commented:
> would you help fix the lint..

Yes, I'll fix this. If I have understood it, I should have used `gofmt -w yourcode.go`. Leason learnt.

tomfotherby commented:
I pushed a fix which I hope will pass the linter :crossed_fingers: . I also added more unit tests to gain more confidence the code works as expected and refactored given the suggestion in https://github.com/kubernetes-sigs/aws-alb-ingress-controller/pull/1162#discussion_r379190396

M00nF1sh commented:
thanks 🤣 
/approve
/lgtm

tdmalone commented:
Thanks so much for this @tomfotherby and for merging it @M00nF1sh. We're really keen to use this and just wondering if there is a target date for the next release containing it? Thanks!

AndrewBoklashko commented:
Hi @tomfotherby @M00nF1sh

I just tried v1.1.6 with this fix. Is this intended to work across multiple hosts? For example:
``` yaml
spec:
  rules:
    - http:
        paths:
          - path: /*
            backend:
              serviceName: ssl-redirect
              servicePort: use-annotation
    - host: hostA
      http:
        paths:
          - path: /pathA/*
            backend:
              serviceName: serviceA
              servicePort: 80
    - host: hostB
      http:
        paths:      
          - path: /*
            backend:
              serviceName: serviceB
              servicePort: 80
```

I believe that if the redirect rule does not have host specified, it should be considered as unconditional redirect as well. So in this case `hostA` and `hostB` rules should be skipped.

tomfotherby commented:
@AndrewBoklashko - Good point. I agree with you. Alas, currently the controller processes each rule in turn and the short circuit mechanism that avoids unneeded rules only operates in the scope of the individual rule.

In my understanding, but please correct me if I'm wrong, the bug you found isn't causing active harm, i.e. it's not causing needed rules to go missing, but it's not operating as well as it could (i.e. it's not doing what it says on the tin). Is that correct?

---

I appologise for this mistake. I hadn't considered "hostless rules" even though I see [the example](https://github.com/kubernetes-sigs/aws-alb-ingress-controller/blob/master/docs/guide/tasks/ssl_redirect.md) does not have a host specified. Thanks for bringing it up - I will work on a patch because we would also benefit from this. I can't promise speed though, I think it will be a level-up in complexity from the current solution and there's rather a lot going on presently...

AndrewBoklashko commented:
> In my understanding, but please correct me if I'm wrong, the bug you found isn't causing active harm, i.e. it's not causing needed rules to go missing, but it's not operating as well as it could (i.e. it's not doing what it says on the tin). Is that correct?

That is correct.

> I appologise for this mistake. I hadn't considered "hostless rules" even though I see the example does not have a host specified. Thanks for bringing it up - I will work on a patch because we would also benefit from this. I can't promise speed though, I think it will be a level-up in complexity from the current solution and there's rather a lot going on presently...

All good, thanks for the follow up. Will be looking to try a patch, but no rush - we can live with how it works now too.

tdmalone commented:
hi @tomfotherby, @AndrewBoklashko, I just thought I'd follow up and see if either you got any further on the above ^^

Is this being tracked under #853? If not, would you like me to create a new issue for it?

tomfotherby commented:
@tdmalone - At my company we don't currently use "hostless rules" in our ALB ingress spec. So to set expectations, I'm not going to be working on an enhancement for the foreseeable future. I would find it hard to test things without a real use-case in my aws account. My PR unfortunately had a bug (raised in https://github.com/kubernetes-sigs/aws-alb-ingress-controller/issues/1274 and fixed in https://github.com/kubernetes-sigs/aws-alb-ingress-controller/pull/1286) which revealed to me the sharp edges and how important it would be to test well. With any luck, other folk will get involved.

> Is this being tracked under #853? If not, would you like me to create a new issue for it?

I would say #853 tracks it.

tdmalone commented:
No worries @tomfotherby, thanks for the reply. I might see if I can refactor to hostless rules and/or try to dig into the code here to see if I can make something work (not a Go programmer, but I could take a stab!)

gpolicante commented:
resolved to version 1.16 https://github.com/kubernetes-sigs/aws-alb-ingress-controller/issues/1274
